### PR TITLE
[BEAM-626] Changes in AvroCoder serialization so it can serialize in Kryo

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:
- [x] Make sure the PR title is formatted like:
  `[BEAM-<Jira issue #>] Description of pull request`
- [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
     Travis-CI on your fork and ensure the whole test matrix passes).
- [x] Replace `<Jira issue #>` in the title with the actual Jira issue
     number, if there is one.
- [ ] If this contribution is large, please file an Apache
     [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).
---

Changes in AvroCoder serialization so it can serialize in Kryo.

Changed writeReplace and readResolve to lazy initialized members.

### CODE REVIEW

**Comment by aviemzur** [comment]
Followed serialization of `ProtoCoder`. See ticket for more details.

**Comment by aviemzur** [comment]
R: @bjchambers or @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java [hunk 1]
```diff
@@ -321,38 +298,79 @@ public void verifyDeterministic() throws NonDeterministicException {
...
   public Schema getSchema() {
-    return schema;
+    return getMemoizedSchema();
+  }
+
+  /**
+   * Get the memoized {@link BinaryDecoder}, possibly initializing it lazily.
+   */
+  private ThreadLocal<BinaryDecoder> getDecoder() {
+    if (memoizedDecoder == null) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java]: this is no longer threadsafe.

Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java]: Do you mean the instantiation of memoizedDecoder should be synchronized so only one is created? I went by what was done in ProtoCoder, but synchronization can be added here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java [hunk 2]
```diff
@@ -174,36 +175,16 @@
...
   // these are ThreadLocal. This code does not need to be re-entrant as AvroCoder does not use
   // an inner coder.
-  private final transient ThreadLocal<BinaryDecoder> decoder;
-  private final transient ThreadLocal<BinaryEncoder> encoder;
-  private final transient ThreadLocal<DatumWriter<T>> writer;
-  private final transient ThreadLocal<DatumReader<T>> reader;
+  private transient ThreadLocal<BinaryDecoder> memoizedDecoder;
+  private transient ThreadLocal<BinaryEncoder> memoizedEncoder;
+  private transient ThreadLocal<DatumWriter<T>> memoizedWriter;
+  private transient ThreadLocal<DatumReader<T>> memoizedReader;
```
Comment (gauravgopi123) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java]: Isn't no-arg constructor needed for this class to work with Kryo serialization and de-serialization

**Comment by gauravgopi123** [comment]
Can desired result not be achieved by just annotating SerializedAvroCoderProxy and AvroCoder classes with `com.esotericsoftware.kryo.DefaultSerializer(com.esotericsoftware.kryo.serializers.JavaSerializer.class)` annotation?

**Comment by aviemzur** [comment]
@gauravgopi123 that would require adding a dependency on kryo in beam-sdks-java-core and adding kryo specific code to AvroCoder.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java (hunk 2 above)
Comment (aviemzur) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java]: @gauravgopi123 There is a strategy in Kryo to serialize classes without no-arg constructors, it is already in use in spark-runner there are classes that require serialization in beam which do not have no-arg constructors.

**Comment by dhalperi** [comment]
Sorry for letting this slip, thanks @amitsela for pinging me.

Do you have a unit test that fails at master but passes here? I think that would give a lot of clarity on the need for this change, and also let me play around to see if I can find a solution I'm happier with :)

**Comment by aviemzur** [comment]
@dhalperi thanks for getting back to me. 
I've added a unit test which fails on master and passes in this PR as you requested. Please see: `org.apache.beam.sdk.coders.AvroCoderTest#testKryoSerialization`
This change required adding a dependency on kryo in beam-sdks-java-core (test scope)

**Comment by amitsela** [comment]
Ping @dhalperi - now that #1143 is merged, this issue becomes pressing as reading from an `UnboundedSource` with the `AvroCoder` will break.
Any news here ? thanks!

**Comment by dhalperi** [comment]
It looks like Kryo should be able to detect the `writeReplace` method and do the right thing automatically. https://github.com/EsotericSoftware/kryo/blob/master/src/com/esotericsoftware/kryo/serializers/ExternalizableSerializer.java#L147

Is there a reason not to enable that for Spark's use of Kryo?

**Comment by dhalperi** [comment]
(Because I can't read javadoc, that's why)

**Comment by dhalperi** [comment]
Okay, I spent a few hours on this, plus @davorbonaci helped :).

A few summaries:

# Not threadsafe now

The current lazy getters are not threadsafe. Making them threadsafe requires some work, see link below.

* I am generally unhappy with putting any synchronized code on the encode()/decode() paths, because those are performance-critical. However, we hope is that a single volatile check will not be too bad.

* So if the perf impact is fairly negligible, well, it's probably fine to go ahead. If there is large perf impact, I'd vote for just registering Kryo to use Java serialization for this class. (And, if necessary, rewriting the serialization code in a better way to not need to expose the inner proxy class.)
    (The AvroCoder serialization was already convoluted crap, this is new crap, no change in amount of crap. ;).

You can assess the performance impact using the Avro microbenchmark:  https://github.com/apache/incubator-beam/blob/master/sdks/java/microbenchmarks/src/main/java/org/apache/beam/sdk/microbenchmarks/coders/AvroCoderBenchmark.java

Link: http://jeremymanson.blogspot.com/2008/05/double-checked-locking.html

# Does threadsafe matter?

Suppose the `volatile` variable access on the hotpath affects perf significantly. Do we care if we just use double-checked locking with no threadsafety?

Maybe not. So we create the ThreadLocals a bunch of times, if the instance is short-lived this may not matter much, and if the coder instance is long lived then eventually it will go across enough synchronization barriers that one will win. We will use extra memory, but since nothing is static we should eventually recover.

-Dan

cc: @lukecwik thoughts?

**Comment by amitsela** [comment]
@dhalperi so just to be clear, current implementation in master _is_ thread safe ?

**Comment by dhalperi** [comment]
@amitsela yes, the current implementation is threadsafe to the best of my knowledge.

**Comment by amitsela** [comment]
@aviemzur can you check with microbenchmark (as Dan proposed) if the current synchronized implementation degrades performance ?
If so, it's best to change this PR to revolve around registration of the AvroCoder with JavaSerializatio  for Kryo.

**Comment by kennknowles** [comment]
I don't have this paged in, and neither the bug nor this PR make clear to me why `AvroCoder` is using Java serialization at all. Given the to/from methods `@JsonCreator` and `asCloudObject`, it should be Jackson-JSON-serialized like all other non-`CustomCoder`s.

FWIW serialization proxies like the one in `AvroCoder` are a standard way to correctly use Java's serialization APIs when you actually design your serialization somewhat. It is pretty surprising that the Kryo/Spark community hasn't been bitten repeatedly, if it is just ignored.

Perhaps everyone just choose between (1) use Java serialization if you don't want to design your serialization (lots of pitfalls, but OK for transmission-only uses) or when you outgrow that (2) use something else entirely. Certainly that's been my trajectory in the past...

We need to take path (2) for Beam coders. We may support language-specific coders in some ways, but they will be by far more costly and non-portable. `AvroCoder` must not be one of them, so it needs to just have a JSON format.

**Comment by lukecwik** [comment]
Ok, I took a stab at this and was able to maintain the existing performance and thread safe like we had before and make the serialization cross compatible with Kryo and Java. Please either merge my commits (https://github.com/lukecwik/incubator-beam/tree/fix941) on top of your existing PR here or close this PR and I can open a new one for review based upon what I have.

@aviemzur What do you think?

**Comment by aviemzur** [comment]
@lukecwik I'll go ahead and close my PR.

**Pull request closed without merging**